File: rdv.py
#!/usr/bin/env python

# RDV Server

import socket
import struct
import sys

def addr2bytes( addr ):
    """Convert an address pair to a hash."""
    host, port = addr
    try:
        host = socket.gethostbyname( host )
    except (socket.gaierror, socket.error):
        raise ValueError, 'invalid host'
    try:
        port = int(port)
    except ValueError:
        raise ValueError, 'invalid port'
    bytes  = socket.inet_aton( host )
    bytes += struct.pack( 'H', port )
    return bytes

def main():
    port = 4653
    try:
        port = int(sys.argv[1])
    except (IndexError, ValueError):
        pass

    sockfd = socket.socket( socket.AF_INET, socket.SOCK_DGRAM )
    sockfd.bind( ('', port) )
    print 'listening on *:%d (udp)' % port

    poolqueue = {}
    while True:
        data, addr = sockfd.recvfrom(32)
        print 'connection from %s:%d' % addr

        pool = data.strip()
        sockfd.sendto( 'ok '+pool, addr )
        data, addr = sockfd.recvfrom(2)
        if data != 'ok':
            continue

        print 'request received for pool:', pool

        try:
            a, b = poolqueue[pool], addr
            sockfd.sendto( addr2bytes(a), b )
            sockfd.sendto( addr2bytes(b), a )
            print 'linked', pool
            del poolqueue[pool]
        except KeyError:
            poolqueue[pool] = addr

if __name__ == '__main__':
    main()

# vim: expandtab shiftwidth=4 softtabstop=4 textwidth=79:

File: nodejs/rdv.js
'use strict'

const dgram = require('dgram')

const port = process.argv[2] || 4653
const sockfd = dgram.createSocket('udp4')

let connections = {}

let poolqueue = {}

sockfd.on('message', (buf, rinfo) => {
  const msg = buf.toString('utf8')

  if (msg === 'ok') {
    const pool = connections[`${rinfo.address}:${rinfo.port}`]
    console.log(`request received for pool: ${pool}`)
    if (poolqueue[pool]) {
      const a = {address: poolqueue[pool].address, port: poolqueue[pool].port}
      const b = {address: rinfo.address, port: rinfo.port}

      sockfd.send(Buffer.from(JSON.stringify(a)), b.port, b.address, (err) => {
        if (err) {
          console.error(`error sending '${JSON.stringify(a)}' to ${b.address}:${b.port}`)
        } else {
          console.log(`ok sending '${JSON.stringify(a)}' to ${b.address}:${b.port}`)
        }
      })

      sockfd.send(Buffer.from(JSON.stringify(b)), a.port, a.address, (err) => {
        if (err) {
          console.error(`error sending '${JSON.stringify(b)}' to ${a.address}:${a.port}`)
        } else {
          console.error(`ok sending '${JSON.stringify(a)}' to ${b.address}:${b.port}`)
        }
      })

      console.log(`linked ${pool}`)
      delete poolqueue[pool]
    } else {
      poolqueue[pool] = rinfo
      delete connections[`${rinfo.address}:${rinfo.port}`]
    }
  } else {
    console.log(`connection from ${rinfo.address}:${rinfo.port}`)

    const pool = msg.replace(/ /g, '').replace(/\s/g, '')

    sockfd.send(Buffer.from(`ok ${pool}`), rinfo.port, rinfo.address, (err) => {
      if (err) {
        console.error(`error sending 'ok ${pool}' to ${rinfo.address}:${rinfo.port}`)
      } else {
        console.log(`ok sending 'ok ${pool}' to ${rinfo.address}:${rinfo.port}`)
        connections[`${rinfo.address}:${rinfo.port}`] = pool
      }
    })
  }
})

sockfd.bind(port, () => {
  console.log(`listening on *:${port} (udp)`)
})
